source/MaterialXRuntime/Private/PvtObject.h

```diff
@@ -258,16 +205,122 @@ class PvtObject : public RtRefCounted<PvtObject>
     TypeBits _typeBits;
     RtToken _name; // TODO: Store a path instead of name token
     PvtPrim* _parent;
-    RtTokenMap<RtTypedValue> _metadataMap;
-    vector<RtToken> _metadataOrder;
+    RtTokenMap<RtTypedValue*> _attr;
+    RtTokenVec _attrNames;
 
     friend class PvtPrim;
-    friend class PvtAttribute;
+    friend class PvtPort;
     friend class PvtInput;
     friend class PvtOutput;
+    friend class PvtNodeGraphPrim;
+    friend class PvtObjectList; 
+    friend class RtAttributeIterator;
     RT_FRIEND_REF_PTR_FUNCTIONS(PvtObject)
 };
 
+
+using PvtObjectVec = vector<PvtObject*>;
+
+// An object container with support for random access, 
+// ordered access and access by name search.
+class PvtObjectList
+{
+public:
+    size_t size() const
+    {
+        return _vec.size();
+    }
+
+    bool empty() const
+    {
+        return _vec.empty();
+    }
+
+    size_t count(const RtToken& name) const
+    {
+        return _map.count(name);
+    }
+
+    PvtObject* find(const RtToken& name) const
+    {
+        auto it = _map.find(name);
+        return it != _map.end() ? it->second.get() : nullptr;
+    }
+
+    PvtObject* operator[](size_t i) const
+    {
+        return i < _vec.size() ? _vec[i] : nullptr;
+    }
+
+    void add(PvtObject* obj)
+    {
+        _map[obj->getName()] = obj->hnd();
+        _vec.push_back(obj);
+    }
+
+    PvtDataHandle remove(const RtToken& name)
+    {
+        auto it = _map.find(name);
+        if (it != _map.end())
+        {
+            PvtDataHandle hnd = it->second;
+            _map.erase(it);
+
+            for (auto it2 = _vec.begin(); it2 != _vec.end(); ++it2)
+            {
+                if (*it2 == hnd.get())
+                {
+                    _vec.erase(it2);
+                    break;
+                }
+            }
+
+            // Return the handled to keep the object alive
+            // if the intent is not to destroy it here.
+            return hnd;
+        }
+        return PvtDataHandle();
+    }
+
+    void rename(const RtToken& name, const RtToken& newName)
+    {
+        auto it = _map.find(name);
+        if (it == _map.end())
+        {
+            throw ExceptionRuntimeError("No object named '" + name.str() + "' exists, unable to rename.");
+        }
+        auto it2 = _map.find(newName);
+        if (newName != name && it2 != _map.end())
+        {
+            throw ExceptionRuntimeError("Another object named '" + newName.str() + "' already exists, unable to rename.");
+        }
+        PvtDataHandle hnd = it->second;
+        _map.erase(it);
+        hnd->setName(newName);
+        _map[newName] = hnd;
+    }
+
+    void clear()
+    {
+        _map.clear();
+        _vec.clear();
+    }
+
+    const PvtObjectVec& vec() const
+    {
+        return _vec;
+    }
+
+    PvtObjectVec& vec()
```

Comment:
Isn't this routine dangerous? Can't users add to the vector and not the map with it?